### PR TITLE
[SPARK-46874][PYTHON] Remove `pyspark.pandas` dependency from `assertDataFrameEqual`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to remove `pyspark.pandas` dependency from `assertDataFrameEqual`

### Why are the changes needed?

To allow `assertDataFrameEqual` when pandas is not installed.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

CI should pass, and manually testing

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/testing/utils.py [hunk 1]
```diff
@@ -758,11 +758,14 @@ def assertDataFrameEqual(
...
         # If pandas dependencies are available, allow pandas or pandas-on-Spark DataFrame
-        import pyspark.pandas as ps
         import pandas as pd
         from pyspark.testing.pandasutils import PandasOnSparkTestUtils
 
         has_pandas = True
+        if has_pandas:
+            # Only import pyspark.pandas when Pandas is available to prevent test stopping
+            # when Pandas is not installed.
+            import pyspark.pandas as ps
```
Comment (HyukjinKwon) [inline python/pyspark/testing/utils.py:768]: ps reference below will fail. Can we work around

Comment (itholic) [reply python/pyspark/testing/utils.py]: relocated the import

**Review by allisonwang-db — APPROVED** [review]
Thanks for the fix

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Could you check Python linter failure, @itholic ?
```
black checks failed:
would reformat /__w/spark/spark/python/pyspark/testing/utils.py

Oh no! 💥 💔 💥
1 file would be reformatted, 976 files would be left unchanged.
Please run 'dev/reformat-python' script.
```

File: python/pyspark/testing/utils.py [hunk 2]
```diff
@@ -768,6 +767,8 @@ def assertDataFrameEqual(
         pass
 
     if has_pandas:
+        import pyspark.pandas as ps
+
```
Comment (panbingkun) [inline python/pyspark/testing/utils.py:771]: From the verification results, there may still be issues:
https://github.com/panbingkun/spark/actions/runs/7684778955/job/20941679265#step:12:4327
image

Comment (dongjoon-hyun) [reply python/pyspark/testing/utils.py:771]: Thank you for reporting, @panbingkun .

To @itholic , is the above failure newly introduced by this PR?

Comment (HyukjinKwon) [reply python/pyspark/testing/utils.py:771]: I think it is not, but would be great to fix it here because essentially the issue still persists.

Comment (itholic) [reply python/pyspark/testing/utils.py:779]: Just fixed it.

BTW, your test environment appears to have `pandas` installed, only `pyarrow` is not installed. Am I understand correctly?? I thought it does not have both `pandas` and `pyarrow`.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/testing/utils.py (hunk 2 above)
Comment (panbingkun) [reply python/pyspark/testing/utils.py:779]: I made some slight improvements, and the test was okay.
https://github.com/apache/spark/pull/44778/files#diff-89e8f0a4413730d86ed98f11e10e3dfe259aa80378a2734eee659ea08bfff8b4R773-R784

**Review by panbingkun — APPROVED** [review]

File: python/pyspark/testing/utils.py (hunk 2 above)
Comment (panbingkun) [reply python/pyspark/testing/utils.py:779]: At present, this changes is okay, and I have already tested it.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you, @itholic and all. Then, let's merge this.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0

**Comment by panbingkun** [comment]
> Thank you, @itholic and all. Then, let's merge this.

In https://github.com/apache/spark/pull/44778, I have merged this PR and let me to verify it again on GA.

**Comment by itholic** [comment]
Thanks for the review @HyukjinKwon @dongjoon-hyun @panbingkun !

**Pull request closed without merging**